### ISSUE TITLE
NonInvariantPropertyType with php_user_filter::$filtername

### ISSUE THREAD

**Issue by BenMorel** [vimeo/psalm#7170]
Consider the following sample filter:

```php
class SkipFilter extends \php_user_filter
{
    public string $filtername;

    public function filter($in, $out, &$consumed, bool $closing): int
    {
        return PSFS_PASS_ON;
    }

    public function onCreate(): bool
    {
        return true;
    }
}
```

Psalm [complains](https://psalm.dev/r/eb30a3ef39):

> ERROR: NonInvariantPropertyType - 8:19 - Property SkipFilter::$filtername has type string, not invariant with php_user_filter::$filtername of type <empty>

However, the property type `string` for `$filtername` is actually required since PHP 8.1:

- [without property type](https://3v4l.org/f2aWh): works on PHP 8.0, fails on 8.1
- [with property type](https://3v4l.org/EmL8A): works on PHP 8.1, fails on 8.0

So, it looks like Psalm's behaviour is currently aligned with PHP 8.0. Would it be possible to use different signatures for `php_user_filter` depending on the current PHP version?

**Comment by AndrolGenhald** [comment]
Indeed, if you wish to do this yourself you can stub the old class somewhere (not sure which stub file would be correct), and add the 8.1 stub to stubs/Php81.phpstub.

**Comment by orklah** [comment]
Seems like we have no particular source of types for php_user_filter. it must comes from reflection, so Psalm behaviour will depend on the php version used to run it.

This is not a good thing and as @AndrolGenhald we need a stub for it. I think we should start with a string docblock on the property and see how it goes with CI. I'm afraid adding the type in signature will make Psalm enforce this on early versions

**Comment by weirdan** [comment]
Actually Psalm knows nothing about `php_user_filter` properties:
```console
$ rg -i 'php_user_filter'

dictionaries/CallMap.php
10157:'php_user_filter::filter' => ['int', 'in'=>'resource', 'out'=>'resource', '&rw_consumed'=>'int', 'closing'=>'bool'],
10158:'php_user_filter::onClose' => ['void'],
10159:'php_user_filter::onCreate' => ['bool'],

dictionaries/CallMap_historical.php
14350:    'php_user_filter::filter' => ['int', 'in'=>'resource', 'out'=>'resource', '&rw_consumed'=>'int', 'closing'=>'bool'],
14351:    'php_user_filter::onClose' => ['void'],
14352:    'php_user_filter::onCreate' => ['bool'],
```

**Comment by weirdan** [comment]
We now have it in our property map, but it doesn't seem to fix the issue.

https://github.com/vimeo/psalm/blob/ceaea625f30289850c8a9274b9c79e21fa848dea/dictionaries/PropertyMap.php#L370-L374
